common/src/main/java/org/apache/rocketmq/common/client/ClientManagerImpl.java

```diff
@@ -0,0 +1,217 @@
+            if (remotingChannel instanceof NettyChannelHandlerContextImpl) {
+                remotingChannel = new NettyChannelImpl(((NettyChannelHandlerContextImpl) remotingChannel).getChannelHandlerContext().channel());
+            }
+            Client oldClient = channelTable.get(remotingChannel);
+            if (oldClient == null) {
+                Client prev = channelTable.put(remotingChannel, client);
+                if (prev != null) {
+                    log.info("New client connected, group: {} {} {} channel: {}", groupId,
+                        client.toString());
+                    updated = true;
+                }
+                oldClient = client;
+            } else {
+                if (!oldClient.getClientId().equals(client.getClientId())) {
+                    log.error(
+                        "[BUG] client channel exist in snode, but clientId not equal. GROUP: {} OLD: {} NEW: {} ",
+                        groupId,
+                        oldClient.toString(),
+                        channelTable.toString());
+                    channelTable.put(remotingChannel, client);
+                }
+            }
+            oldClient.setLastUpdateTimestamp(System.currentTimeMillis());
+            onRegister(groupId, remotingChannel);
+        }
+        log.debug("Register client role: {}, group: {}, last: {}", client.getClientRole(), groupId,
+            client.getLastUpdateTimestamp());
+        return updated;
+    }
+
+    protected void removeClient(String groupId, RemotingChannel remotingChannel) {
+        ConcurrentHashMap<RemotingChannel, Client> channelTable = groupClientTable.get(groupId);
+        if (channelTable != null) {
+            Client prev = channelTable.remove(remotingChannel);
+            if (prev != null) {
+                log.info("Unregister client: {}  in  group, {}", prev, groupId);
+            }
+            if (channelTable.isEmpty()) {
+                groupClientTable.remove(groupId);
+                log.info("Unregister client ok, no any connection, and remove consumer group, {}",
+                    groupId);
+            }
+        }
+    }
+
+    @Override
+    public void unRegister(String groupId, RemotingChannel remotingChannel) {
+        removeClient(groupId, remotingChannel);
+        onUnregister(groupId, remotingChannel);
+
+    }
+
+    @Override
+    public void onClose(Set<String> groups, RemotingChannel remotingChannel) {
+        for (String groupId : groups) {
+            removeClient(groupId, remotingChannel);
+            onClosed(groupId, remotingChannel);
+        }
+    }
+
+    public List<RemotingChannel> getChannels(String groupId) {
+        if (groupId != null) {
+            List<RemotingChannel> result = new ArrayList<RemotingChannel>();
+            ConcurrentHashMap channelsMap = this.groupClientTable.get(groupId);
+            if (channelsMap != null) {
+                result.addAll(this.groupClientTable.get(groupId).keySet());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `get()` may not be thread-safe at lines: **178, 179, and 180**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The remove that can remove the key is at line: **153**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).